### ISSUE TITLE
Configurable Product Indexing Issue Magento 2.2.6

### ISSUE THREAD

**Issue by gavinlimely** [magento/magento2#19407]
### Preconditions (*)
1. Magento 2.2.6

### Steps to reproduce (*)
1. catalog_product_price index reindexes configurable product prices incorrectly. This is triggered by a cron task.
2. This only happens when the process is triggered by the cron. When performing a full re-index, this bug does not exist.

### Expected result (*)
1. Configurable products have min_price & max_price indexed correctly.

### Actual result (*)
1. Most configurable product prices are changed to have a min_price & max_price set to 0 within the catalog_product_index_price table.

I've done some research & somebody has detailed the problem in this post: 

https://magento.stackexchange.com/questions/242801/magento-2-2-6-possible-problems#answer-244821

I'll post it here for convenience too : 

"I've encountered 2 problems regarding price reindexing. One problem is that configurable price indexer (and probably other complex products too but I have no production project who uses bundles and grouped products extensively) relies on its child product index to be properly executed. Configurable price indexer works in 2 step:

build index data based on configurable product price which is null as magento does not allow you to set price on configurable price level
update min and max columns in catalog_product_index_price table based on children data from the same table
The first bug I found was that reindexList method which is fired by scheduled reindex reindexed the given list by product type which was sorted based on ids provided. If the lowest id in the list was simple then all simples were reindexed before configurables, if it was configurable then configurables before simples.

The second bug relied on the fact that in the first step of the configurable price reindex process table catalog_product_index_price was updated based on a query which created where part as where e.entity_id BEETWEEN ? and ? and values where min and max ids from the provided list. This resulted in index table being updated for all configurables form the range based on its data and in most cases all price column were set to 0. In the second step, where indexer should update min and max price based on children query was executed with where part like where le.entity_id IN (?), so only for the configurables provided to the reindexList method making all other in the base state."

**Comment by magento-engcom-team** [comment]
Hi @gavinlimely. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@gavinlimely do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-andrii. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by ghost** [comment]
Hi @gavinlimely. Thank you for your report. This issue is a duplicate https://github.com/magento/magento2/issues/7367

**Comment by perryholden** [comment]
This is not a duplicate of 7367. I very recently created this composer patch for v 2.2.6. @gavinlimely - if you are still experiencing this, can you apply the following and see if it fixes your issue?

```
diff --git a/Model/ResourceModel/Product/Indexer/Price/Configurable.php b/Model/ResourceModel/Product/Indexer/Price/Configurable.php
index 81e2e99..59cc9d9 100644
--- a/Model/ResourceModel/Product/Indexer/Price/Configurable.php
+++ b/Model/ResourceModel/Product/Indexer/Price/Configurable.php
@@ -197,7 +197,8 @@ class Configurable implements DimensionalIndexerInterface
             ['le.entity_id', 'customer_group_id', 'website_id']
         );
         if ($entityIds !== null) {
-            $select->where('le.entity_id IN (?)', $entityIds);
+            $select->where(sprintf('le.entity_id BETWEEN %s AND %s', min($entityIds), max($entityIds)));
+            $select->where("le.type_id = ?", \Magento\ConfigurableProduct\Model\Product\Type\Configurable::TYPE_CODE);
         }
         $query = $select->insertFromSelect($temporaryOptionsTableName);
         $this->getConnection()->query($query);
```

**Comment by jefflenz** [comment]
Happening on 2.3.2-p2 enterprise

**Comment by aimiliano** [comment]
hello,
we experience the same issue exactly as you describe it in m2 ee 2.3.2. have you found any solution in the problem?

**Comment by joshiprashiddha** [comment]
Followed @dewayneholden suggestion, that solved my issue for CE 2.2.6
Checked same line in CE 2.3.2, its still not corrected.

**Comment by theloveofcode** [comment]
Same exact issue on 2.3.2-p2 enterprise as well.

**Comment by engcom-Charlie** [comment]
Hello @gavinlimely 
Thank you for contribution and collaboration!

We are not able to reproduce this issue on the latest 2.4-develop branch by provided steps.

**Testing scenario:**

1. Check configurable product data in `catalog_product_index_price ` table 
![1](https://user-images.githubusercontent.com/51680850/70703047-f5773c80-1cd7-11ea-9f97-58e5f77d4b6a.png)

2. Create a new catalog product price for all products and apply it
![2](https://user-images.githubusercontent.com/51680850/70703077-fe680e00-1cd7-11ea-8638-85df95e82a46.png)

3. Wait for cron re-index
4. Check configurable product data in `catalog_product_index_price ` table again

**Result:**
Configurable products have min_price & max_price indexed correctly:
![3](https://user-images.githubusercontent.com/51680850/70703100-0fb11a80-1cd8-11ea-9ce0-eacaadf1bd6d.png)

We are closing this issue due to branch `2.2-develop` was closed and
> Magento no longer accepts pull requests and issues to the v2.2 release lines to focus all development efforts on v2.3.

 See [Accepted pull requests and ported code](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#rules) for more details

If you still faced this issue on `2.3 - 2.4` please create a new Issue with all required details according to [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines)

Thanks for your report!
